### PR TITLE
EntityManager::__cosntruct() as public method

### PR DESCRIPTION
**Pull request by mcspronko**
Updated EntityManager::__construct() method to public visibility:
- moved create() method to EntityManagerFactory class
- added new Configuration class methods for UnitOfWork and ProxyFactory classes
- updated documentation
- updated unit and functional tests

### CODE REVIEW

File: lib/Doctrine/ORM/EntityManager.php [hunk 1]
```diff
@@ -813,44 +816,6 @@ public function initializeObject($obj)
...
-     * @param mixed         $conn         An array with the connection parameters or an existing Connection instance.
-     * @param Configuration $config       The Configuration instance to use.
-     * @param EventManager  $eventManager The EventManager instance to use.
-     *
-     * @return EntityManager The created EntityManager.
-     *
-     * @throws \InvalidArgumentException
-     * @throws ORMException
-     */
-    public static function create($conn, Configuration $config, EventManager $eventManager = null)
```
Comment (fprochazka) [inline lib/Doctrine/ORM/EntityManager.php]: removing this method would an unnecesary BC Break

File: lib/Doctrine/ORM/EntityManager.php [hunk 2]
```diff
@@ -157,8 +156,12 @@ protected function __construct(Connection $conn, Configuration $config, EventMan
         $this->metadataFactory->setCacheDriver($this->config->getMetadataCacheImpl());
 
         $this->repositoryFactory = $config->getRepositoryFactory();
-        $this->unitOfWork        = new UnitOfWork($this);
-        $this->proxyFactory      = new ProxyFactory(
+
+        $unitOfWorkClassName  = $config->getUnitOfWorkClassName();
+        $this->unitOfWork     = new $unitOfWorkClassName($this);
+
+       $proxyFactoryClassName = $config->getProxyFactoryClassName();
```
Comment (fprochazka) [inline lib/Doctrine/ORM/EntityManager.php]: your whitespaces are a little off

File: lib/Doctrine/ORM/EntityManager.php (hunk 1 above)
Comment (stekycz) [inline lib/Doctrine/ORM/EntityManager.php]: agree

Comment (mcspronko) [inline lib/Doctrine/ORM/EntityManager.php]: As for now EntityManager has a lot of responsibilities including creation of its own instance. Separate class should prepare and instantiate EntityManager. I agree that this is backward incompatible change. The strategy might be changed based on release policy. In case 2.5.* allows to introduce this change, we are in a good shape. Alternatively, this method might marked as @deprecated since 2.5 and proxy behavior to the EntityManagerFactory
Like this:
{code}
    /**
     * @deprecated since 2.5
     */
    public static function create($conn, Configuration $config, EventManager $eventManager = null)
    {
        return new EntityManagerFactory($conn, $config, $eventManager);
    }
{code}

Thanks

File: lib/Doctrine/ORM/EntityManager.php [hunk 3]
```diff
@@ -157,8 +156,12 @@ protected function __construct(Connection $conn, Configuration $config, EventMan
         $this->metadataFactory->setCacheDriver($this->config->getMetadataCacheImpl());
 
         $this->repositoryFactory = $config->getRepositoryFactory();
-        $this->unitOfWork        = new UnitOfWork($this);
-        $this->proxyFactory      = new ProxyFactory(
+
+        $unitOfWorkClassName  = $config->getUnitOfWorkClassName();
+        $this->unitOfWork     = new $unitOfWorkClassName($this);
```
Comment (stof) [inline lib/Doctrine/ORM/EntityManager.php]: please don't do this. Configuring a class name is a bad extension point, because it assumes what the dependencies of the class are. If we make it configurable, we should do it cleanly. thus, this is a totally different topic than making ``EntityManager::__construct`` public, so it should be submitted separately (I'm against making the UnitOfWork class configurable btw, so I would vote for rejecting this PR entirely because of this change given that you are mixing things)

File: lib/Doctrine/ORM/EntityManager.php (hunk 1 above)
Comment (stof) [inline lib/Doctrine/ORM/EntityManager.php]: @mcspronko BC breaks are forbidden in minor versions. Semver means that any BC break in the public API requires bumping the major version

File: lib/Doctrine/ORM/EntityManager.php (hunk 3 above)
Comment (mcspronko) [inline lib/Doctrine/ORM/EntityManager.php]: @stof the reason behind configuring UnitOfWork class name is that next layer (UnitOfWOrk) depends on higher level (EntityManager). So refactoring of these 2 layers would be a really major one.

File: lib/Doctrine/ORM/EntityManager.php (hunk 1 above)
Comment (mcspronko) [inline lib/Doctrine/ORM/EntityManager.php]: @stof so option with deprecated create() method should be ok

Comment (fprochazka) [inline lib/Doctrine/ORM/EntityManager.php]: You can mark it deprecated, but you cannot change the behaviour.

Comment (mcspronko) [inline lib/Doctrine/ORM/EntityManager.php]: Agree, my mistake it should be like this
[code]
    /**
     * @deprecated since 2.5
     */
    public static function create($conn, Configuration $config, EventManager $eventManager = null)
    {
        return new EntityManagerFactory::create($conn, $config, $eventManager);
    }
[code]

where create() behavior is the same.
Thanks

Comment (fprochazka) [inline lib/Doctrine/ORM/EntityManager.php]: why do you have the `new` there?

Comment (mcspronko) [inline lib/Doctrine/ORM/EntityManager.php]: TYPO :)

File: lib/Doctrine/ORM/EntityManager.php (hunk 3 above)
Comment (mcspronko) [inline lib/Doctrine/ORM/EntityManager.php]: @stof when do you think you can contribute configurable UnitOfWork?

Comment (stof) [inline lib/Doctrine/ORM/EntityManager.php]: @mcspronko it should be in a separate PR. And you should explain clearly *why* you need this. The UnitOfwork is quite sensible logic, so I'm not sure it makes sense to make it configurable.

Comment (mcspronko) [inline lib/Doctrine/ORM/EntityManager.php]: @stof not sure I understand where we have sensible logic in UnitOfWork. There are cases where you need to customize UnitOfWork, however having no interfaces force to copy-paste EntityManager, UnitOfWork, ProxyFactory, collections, persisters, hydrators etc.

Comment (stof) [inline lib/Doctrine/ORM/EntityManager.php]: @mcspronko what are your use case for customizing the UnitOfWork ? This is what you need to explain. We won't make it configurable without knowing why (and this is also why it should be in a dedicated PR).
And what is sensible in the UoW is both performance (it has a huge impact on the performance of Doctrine) and reliability (you should not start replacing the UoW implementation before you understand all the logic in the existing one. This will take time. We have lots of sensible logic in it)

**Comment by Majkl578** [comment]
EntityManager::__construct() is not public by design and EntityManager::create() is the only exposed and proper way to instantiate new instance. Also EntityManager will become final in Doctrine 3.0.
Additionally configuring UnitOfWork class name shouldn't be possible, it is not intended to be extended or replaced as it contains pretty sensitive code.

Valid extension point to enhance EntityManager is through decorator pattern using EntityManagerInterface, not inheritance.
If you still think this is needed and want to update this PR (if so, please on top of develop branch), let me know and I will reopen. I think extracting factory to separate class could be eventually considered, but not configurable UnitOfWork. Thanks.
